File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.py
class Trie:

    def __init__(self):
        self.children = {}
        self.isEnd = False

    def insert(self, word: str) -> None:
        cur = self
        for i in range(len(word)):
            if word[i] not in cur.children:
                cur.children[word[i]] = Trie()
            cur = cur.children[word[i]]

        cur.isEnd = True
        
    def search(self, word: str) -> bool:
        cur = self
        for i in range(len(word)):
            if word[i] not in cur.children:
                return False
            cur = cur.children[word[i]]

        return cur.isEnd

    def startsWith(self, prefix: str) -> bool:
        cur = self
        for i in range(len(prefix)):
            if prefix[i] not in cur.children:
                return False
            cur = cur.children[prefix[i]]

        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class Trie {
public:
    unordered_map<char, Trie*> children;
    bool isEnd;
    Trie() {
        isEnd = false;
    }
    
    void insert(string word) {
        Trie* cur = this;
        for (char c: word) {
            if (cur -> children.find(c) == cur -> children.end()) {
                cur -> children[c] = new Trie();
            }
            cur = cur -> children[c];
        }
        cur -> isEnd = true;
    }
    
    bool search(string word) {
        Trie* cur = this;
        for (char c: word) {
            if (cur -> children.find(c) != cur -> children.end()) {
                cur = cur -> children[c];
            } else {
                return false;
            }
        }
        return cur -> isEnd;
    }
    
    bool startsWith(string prefix) {
        Trie* cur = this;
        for (char c: prefix) {
            if (cur -> children.find(c) != cur -> children.end()) {
                cur = cur -> children[c];
            } else {
                return false;
            }
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */